velodrome/token-counter/token-counter.go

```diff
@@ -0,0 +1,209 @@
+	login, err := GetUsername(client) // Get user name for token
+	if err != nil {
+		return nil, err
+	}
+
+	return &TokenHandler{
+		gClient:  client,
+		login:    login,
+		influxdb: influxdb,
+	}, nil
+}
+
+// CreateTokenHandlers goes through the list of token files, and create handlers
+func CreateTokenHandlers(tokenFiles []string, influxdb *InfluxDB) ([]TokenHandler, error) {
+	tokens := []TokenHandler{}
+	for _, tokenFile := range tokenFiles {
+		f, err := os.Open(tokenFile)
+		if err != nil {
+			return nil, fmt.Errorf("Can't open token-file (%s): %s", tokenFile, err)
+		}
+		token, err := CreateTokenHandler(f, influxdb)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create token (%s): %s", tokenFile, err)
+		}
+		tokens = append(tokens, *token)
+	}
+	return tokens, nil
+}
+
+func (t TokenHandler) getCoreRate() (*github.Rate, error) {
+	limits, _, err := t.gClient.RateLimits()
+	if err != nil {
+		return nil, err
+	}
+	return limits.Core, nil
+}
+
+// Process does the main job:
+// It tries to get the value of "Remaining" rate just before the token
+// gets reset. It does that more and more often (as the reset date gets
+// closer) to get the most accurate value.
+func (t TokenHandler) Process() {
+	lastRate, err := t.getCoreRate()
+	if err != nil {
+		glog.Fatalf("%s: Couldn't get rate limits:", t.login, err)
+	}
+
+	for {
+		halfPeriod := lastRate.Reset.Time.Sub(time.Now()) / 2
+		glog.Infof("%s: Current rate: %s. Sleeping for %s.", t.login, lastRate, halfPeriod)
+		time.Sleep(halfPeriod)
+		newRate, err := t.getCoreRate()
+		if err != nil {
+			glog.Error("Failed to get CoreRate: ", err)
+			continue
+		}
+		// There is a bug in Github. They seem to reset the Remaining value before reseting the Reset value.
+		if !newRate.Reset.Time.Equal(lastRate.Reset.Time) || newRate.Remaining > lastRate.Remaining {
+			glog.Infof(
+				"%s: ### TOKEN USAGE: %d",
+				t.login,
+				lastRate.Limit-lastRate.Remaining,
+			)
+			t.influxdb.Push(
+				"github_token_count",
+				map[string]string{"login": t.login},
+				map[string]interface{}{"value": lastRate.Limit - lastRate.Remaining},
+				lastRate.Reset.Time,
+			)
+			// Make sure the timer is properly reset, and we have time anyway
+			time.Sleep(30 * time.Minute)
+			for {
+				newRate, err = t.getCoreRate()
+				if err == nil {
+					break
+				}
+				glog.Error("Failed to get CoreRate: ", err)
+				time.Sleep(time.Minute)
+			}
+
+		}
+		lastRate = newRate
+	}
+}
+
+func runProgram(flags *tokenCounterFlags) error {
+	influxdb, err := flags.influx.CreateDatabase()
+	if err != nil {
+		return err
+	}
+
+	tokens, err := CreateTokenHandlers(flags.tokens, influxdb)
+	if err != nil {
+		return err
+	}
+
+	if len(tokens) == 0 {
+		glog.Warning("No token given, nothing to do. Leaving...")
+		return nil
+	}
+
+	for _, token := range tokens {
+		glog.Infof("Processing token for '%s'", token.login)
+		go token.Process()
+	}
+
+	select {}
```

Comment:
Yes, each goroutine polls rate_limits from github, the main thread does nothing.